### PR TITLE
Make the timestamper do timestamping exactly once per source 

### PR DESCRIPTION
**Pull request by ruchirK**
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5585)

### CODE REVIEW

**Comment by ruchirK** [comment]
@benesch running into a weird thing with this pr: basically I'm changing timestamping to happen on a "once per source" basis instead of a "once per source instance basic" so this moves the communication of "tell the timestamper about a new source to timestamp" from coord -> dataflow-workers -> coord -> timestamper to instead just be: coord -> timestamper.

Unfortunately, on restart, I would like the coord to tell the timestamper about sources to timestamp at bootstrap time, but the timestamper doesn't exist at that point (because the timestamper waits till the internal command mpsc is instantiated in serve)

I can think of three ways to resolve this:

1. have the timestamper publish messages meant for the coord to a different, unique channel - then we can initialize timestamper roughly where the caching thread gets started
2. keep some state about initial sources and connectors in a vector when we bootstrap, and pass that to the timestamper when we initialize it
3. scan the catalog for sources right before we initialize the timestamper, send all of those to the timestamper on init

2 and 3 are not really different. do you have a preference / alternative opinion?

**Comment by benesch** [comment]
Sorry, just getting through my email backlog now. My inclination would be to do something roughly like (1). I think just creating `internal_cmd_tx` and `internal_cmd_rx` in `bootstrap` and passing those into `serve` would do the trick.

**Comment by ruchirK** [comment]
> Sorry, just getting through my email backlog now. My inclination would be to do something roughly like (1). I think just creating `internal_cmd_tx` and `internal_cmd_rx` in `bootstrap` and passing those into `serve` would do the trick.

This makes sense. I'll spin that up in a separate pr that we can hopefully merge today!

File: src/dataflow/src/server.rs [hunk 1]
```diff
@@ -746,18 +760,28 @@ where
...
                     }
-                    let source = self
-                        .render_state
-                        .ts_source_mapping
-                        .get(&id)
-                        .expect("Id should be present");
-                    if let Some(source) = source.upgrade() {
-                        if let Some(token) = &*source {
-                            token.activate();
+                    // TODO: clean this up to not do so much extra work each time
```
Comment (ruchirK) [inline src/dataflow/src/server.rs:763]: Don't forget about this!

File: src/dataflow/src/source/mod.rs [hunk 2]
```diff
@@ -573,15 +546,20 @@ impl ConsistencyInfo {
             // Per partition, we iterate over the data structure to remove (ts,offset) mappings for which
             // we have seen all records <= offset. We keep track of the last "closed" timestamp in that partition
             // in next_partition_ts
-            if let Some(entries) = timestamp_histories.borrow_mut().get_mut(id) {
+            if let Some(entries) = timestamp_histories.borrow().get(&id.source_id) {
```
Comment (ruchirK) [inline src/dataflow/src/source/mod.rs:549]: lets see if we can't clean this up / at least mention that the linear scan for BYO sources is subpoptimal and will be fixed soon

**Comment by ruchirK** [comment]
I think this pr is ready for a real review!

File: src/testdrive/src/action.rs [hunk 3]
```diff
@@ -103,12 +103,14 @@ pub struct State {
 
 impl State {
     pub async fn reset_materialized(&mut self) -> Result<(), Error> {
+        println!("in reset");
```
Comment (ruchirK) [inline src/testdrive/src/action.rs:106]: get rid of these

File: src/coord/src/coord.rs [hunk 4]
```diff
@@ -3288,6 +3273,28 @@ impl Coordinator {
         }
     }
 
+    // Notify the timestamper thread that a source has been created or dropped.
+    async fn update_timestamper(&mut self, source_id: GlobalId, create: bool) {
```
Comment (mjibson) [inline src/coord/src/coord.rs:3277]: The if branches in this function don't share any code and each variant (create = true/false) is only called in one place. This suggests to me this shouldn't be a function? Light preference to just inlining the code.

Comment (ruchirK) [reply src/coord/src/coord.rs:3277]: I think the create branch is called twice? (once when we create the source and once on restart). the drop variant is only called once though

Comment (mjibson) [reply src/coord/src/coord.rs:3277]: Oh right, I missed one. Not opposed to leaving it as is, just a thought I had.

**Review by benesch — APPROVED** [review]
I think this makes sense, though I can't claim to have thought about this in great detail nor reviewed the code with too much scrutiny. But if tests are happy I say let's go for it. Sorry for the delay; I'm pretty swamped at the moment. :(

**Pull request merged**